jigglypuffRL/common/trainer.py

```diff
@@ -0,0 +1,203 @@
+                raise Exception("Off Policy Training requires a Replay Buffer")
+            else:
+                self.buffer = self.agent.replay_buffer
+        self.save_interval = save_interval
+        self.render = render
+        self.max_ep_len = max_ep_len
+        self.ckpt_log_name = ckpt_log_name
+        self.steps_per_epoch = steps_per_epoch
+        self.epochs = epochs
+        self.device = device
+        self.log_interval = log_interval
+        self.batch_size = batch_size
+        if seed is not None:
+            set_seeds(seed, self.env)
+
+    def train(self):
+        raise NotImplementedError
+
+    def save(self):
+        saving_params = self.agent.get_hyperparams()
+        logdir = self.logger.logdir
+        algo = self.agent.__class__.__name__
+        env_name = self.env.envs[0].unwrapped.spec.id
+
+        save_dir = '{}/checkpoints/{}_{}'.format(logdir, algo, env_name)
+        os.makedirs(save_dir, exist_ok=True)
+        torch.save('{}/{}.pt'.format(save_dir, self.ckpt_log_name))
+
+    @property
+    def n_envs(self):
+        return self.env.n_envs
+
+
+class OffPolicyTrainer(Trainer):
+    def __init__(self, agent, env, logger, buffer=None, off_policy=True, save_interval=0,
+                 render=False, max_ep_len=1000, distributed=False, ckpt_log_name='experiment',
+                 steps_per_epoch=4000, epochs=10, device='cpu', log_interval=10, batch_size=50,
+                 warmup_steps=1000, start_update=1000, update_interval=50, seed=0):
+        super(OffPolicyTrainer, self).__init__(agent, env, logger, buffer, 
+                                               off_policy, save_interval, render, max_ep_len,
+                                               distributed, ckpt_log_name,steps_per_epoch, 
+                                               epochs, device, log_interval, batch_size, seed
+                                               )
+        self.warmup_steps = warmup_steps
+        self.update_interval = update_interval
+        self.start_update = start_update
+        
+    def train(self):
+        state, episode_reward, episode_len, episode = self.env.reset(), 0, 0, 0
+        total_steps = self.steps_per_epoch * self.epochs
+
+        if 'noise' in self.agent.__dict__ and self.agent.noise is not None:
+            self.agent.noise.reset()
+
+        for t in range(total_steps):
+
+            if t < self.warmup_steps:
+                action = self.env.action_space.sample()
+            else:
+                action = self.agent.select_action(state)
+
+            next_state, reward, done, info = self.env.step(action)
+            if self.render:
+                self.env.render()
+
+            episode_reward += reward
+            episode_len += 1
+            
+            done = False if episode_len == self.max_ep_len else done
+
+            self.buffer.push((state, action, reward, next_state, done))
+
+            states = next_state
```

Comment:
This is why its not training

Reply:
Hmm. Maybe we just keep a separate function like 'observe' for this initial stage to avoid such errors ahead.

Reply:
On the other hand, trains like a charm now.

Reply:
Yeah I think stuff like this is what Pranav (Garg) meant when he said we should write code based on tests

Reply:
Yeah it might be hard to write tests to check if algos are 'trainable'. We might have to think some hacky way to get around this.